File: fps_overlay.py
"""fps_overlay.py

Draws a live per-camera FPS readout in the top-left corner of each tile in the
debug display (and in any recording), using nvdsosd display metadata.

How it fits together:
  * ``FpsMeter`` measures a rolling frames-per-second for each camera. It is
    ticked once per frame per camera from the detection probe on nvtracker's src
    pad — a point where the batch reliably carries one frame_meta per camera.
  * ``attach_fps_overlay`` adds a second probe on nvmultistreamtiler's src pad
    that, once per composited buffer, writes one text label per camera at that
    camera's tile corner. Measuring before the tiler and drawing after it keeps
    both steps on solid ground (guaranteed per-source frames upstream; a single
    known composite coordinate space downstream).

Only used in --display / --record modes.
"""

from __future__ import annotations

import time

import gi

gi.require_version("Gst", "1.0")
from gi.repository import Gst  # noqa: E402

import pyds  # noqa: E402


class FpsMeter:
    """Rolling per-camera FPS estimate.

    Args:
        window_s: length of the averaging window in seconds. Every ``window_s``
            the fps for a camera is recomputed from the frames counted in that
            window (so the number is responsive but not jittery).
    """

    def __init__(self, window_s: float = 0.5) -> None:
        self._window = float(window_s)
        self._count: dict = {}   # source_id -> frames counted this window
        self._start: dict = {}   # source_id -> window start (monotonic)
        self._fps: dict = {}     # source_id -> last computed fps

    def tick(self, source_id: int) -> None:
        """Record one processed frame for ``source_id`` and refresh its fps."""
        now = time.monotonic()
        self._count[source_id] = self._count.get(source_id, 0) + 1
        start = self._start.setdefault(source_id, now)
        elapsed = now - start
        if elapsed >= self._window:
            self._fps[source_id] = self._count[source_id] / elapsed
            self._count[source_id] = 0
            self._start[source_id] = now

    def get(self, source_id: int) -> float:
        """Return the latest fps for ``source_id`` (0.0 until first window closes)."""
        return self._fps.get(source_id, 0.0)


def attach_fps_overlay(tiler: Gst.Element, num_cams: int, meter: FpsMeter) -> None:
    """Attach a probe on the tiler's src pad that draws per-tile FPS text.

    Args:
        tiler: the ``nvmultistreamtiler`` element (its rows/columns/width/height
            define the tile grid we position text into).
        num_cams: number of cameras (== number of tiles == labels to draw).
        meter: the shared ``FpsMeter`` fed by the detection probe.

    The tiler lays streams out in row-major order by source_id, so camera ``i``
    occupies tile (row = i // columns, col = i % columns). We convert that to the
    composited output's pixel coordinates and drop a label just inside the corner.
    """
    cols = int(tiler.get_property("columns"))
    rows = int(tiler.get_property("rows"))
    out_w = int(tiler.get_property("width"))
    out_h = int(tiler.get_property("height"))
    tile_w = out_w / max(cols, 1)
    tile_h = out_h / max(rows, 1)

    def probe(pad: Gst.Pad, info: Gst.PadProbeInfo, _user) -> Gst.PadProbeReturn:
        buf = info.get_buffer()
        if buf is None:
            return Gst.PadProbeReturn.OK
        batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(buf))
        if batch_meta is None:
            return Gst.PadProbeReturn.OK
        l_frame = batch_meta.frame_meta_list
        if l_frame is None:
            return Gst.PadProbeReturn.OK
        try:
            # Display metadata is per-frame; attach our whole label set to the
            # first frame_meta. text_params x/y_offset are absolute pixels in the
            # composited surface, so one frame_meta can carry every tile's label.
            frame_meta = pyds.NvDsFrameMeta.cast(l_frame.data)
        except StopIteration:
            return Gst.PadProbeReturn.OK

        display_meta = pyds.nvds_acquire_display_meta_from_pool(batch_meta)
        n = 0
        for sid in range(num_cams):
            col = sid % cols
            row = sid // cols
            txt = display_meta.text_params[n]
            txt.display_text = f"cam{sid}  {meter.get(sid):4.1f} FPS"
            txt.x_offset = int(col * tile_w) + 12
            txt.y_offset = int(row * tile_h) + 10
            txt.font_params.font_name = "Serif"
            txt.font_params.font_size = 12
            txt.font_params.font_color.set(1.0, 1.0, 1.0, 1.0)   # white text
            txt.set_bg_clr = 1
            txt.text_bg_clr.set(0.0, 0.0, 0.0, 0.55)             # translucent black box
            n += 1
        display_meta.num_labels = n
        pyds.nvds_add_display_meta_to_frame(frame_meta, display_meta)
        return Gst.PadProbeReturn.OK

    src_pad = tiler.get_static_pad("src")
    if src_pad is None:
        raise RuntimeError("nvmultistreamtiler has no src pad for the FPS overlay probe.")
    src_pad.add_probe(Gst.PadProbeType.BUFFER, probe, None)

File: cpp/src/fps_overlay.cpp
#include "fps_overlay.hpp"

#include <glib.h>

#include <algorithm>
#include <memory>
#include <stdexcept>

#include "gstnvdsmeta.h"
#include "nvdsmeta.h"

namespace mcrt {

namespace {
double mono_secs() { return g_get_monotonic_time() / 1e6; }

struct OverlayCtx {
  FpsMeter* meter;
  int num_cams;
  int cols;
  double tile_w;
  double tile_h;
};

GstPadProbeReturn overlay_probe(GstPad*, GstPadProbeInfo* info,
                                gpointer user_data) {
  auto* ctx = static_cast<OverlayCtx*>(user_data);
  GstBuffer* buf = GST_PAD_PROBE_INFO_BUFFER(info);
  if (buf == nullptr) return GST_PAD_PROBE_OK;
  NvDsBatchMeta* batch_meta = gst_buffer_get_nvds_batch_meta(buf);
  if (batch_meta == nullptr || batch_meta->frame_meta_list == nullptr)
    return GST_PAD_PROBE_OK;

  /* After the tiler the batch is ONE composited frame; text offsets are
   * absolute pixels in that surface, so the first frame_meta can carry every
   * tile's label. */
  auto* frame_meta =
      static_cast<NvDsFrameMeta*>(batch_meta->frame_meta_list->data);
  if (frame_meta == nullptr) return GST_PAD_PROBE_OK;

  NvDsDisplayMeta* display_meta =
      nvds_acquire_display_meta_from_pool(batch_meta);
  int n = 0;
  const int max_labels =
      std::min(ctx->num_cams, static_cast<int>(MAX_ELEMENTS_IN_DISPLAY_META));
  for (int sid = 0; sid < max_labels; ++sid) {
    const int col = sid % ctx->cols;
    const int row = sid / ctx->cols;
    NvOSD_TextParams* txt = &display_meta->text_params[n];
    /* display_text is freed by the OSD meta pool — must be g_malloc'd. */
    txt->display_text =
        g_strdup_printf("cam%d  %4.1f FPS", sid, ctx->meter->get(sid));
    txt->x_offset = static_cast<unsigned int>(col * ctx->tile_w) + 12;
    txt->y_offset = static_cast<unsigned int>(row * ctx->tile_h) + 10;
    txt->font_params.font_name = const_cast<char*>("Serif");
    txt->font_params.font_size = 12;
    txt->font_params.font_color = {1.0, 1.0, 1.0, 1.0};  // white text
    txt->set_bg_clr = 1;
    txt->text_bg_clr = {0.0, 0.0, 0.0, 0.55};            // translucent black
    ++n;
  }
  display_meta->num_labels = n;
  nvds_add_display_meta_to_frame(frame_meta, display_meta);
  return GST_PAD_PROBE_OK;
}
}  // namespace

void FpsMeter::tick(int source_id) {
  const double now = mono_secs();
  std::lock_guard<std::mutex> lock(mu_);
  Slot& s = slots_[source_id];
  if (s.count == 0 && s.start == 0.0) s.start = now;
  s.count += 1;
  const double elapsed = now - s.start;
  if (elapsed >= window_) {
    s.fps = s.count / elapsed;
    s.count = 0;
    s.start = now;
  }
}

double FpsMeter::get(int source_id) {
  std::lock_guard<std::mutex> lock(mu_);
  auto it = slots_.find(source_id);
  return it == slots_.end() ? 0.0 : it->second.fps;
}

void attach_fps_overlay(GstElement* tiler, int num_cams, FpsMeter* meter) {
  gint cols = 1, rows = 1;
  guint out_w = 1280, out_h = 720;
  g_object_get(tiler, "columns", &cols, "rows", &rows, "width", &out_w,
               "height", &out_h, nullptr);

  /* Owned by the probe: the destroy-notify below deletes it when the pad
   * (and with it the probe) is torn down, so no explicit cleanup needed. */
  auto* ctx = new OverlayCtx();
  ctx->meter = meter;
  ctx->num_cams = num_cams;
  ctx->cols = std::max(cols, 1);
  ctx->tile_w = static_cast<double>(out_w) / std::max(cols, 1);
  ctx->tile_h = static_cast<double>(out_h) / std::max(rows, 1);

  GstPad* src_pad = gst_element_get_static_pad(tiler, "src");
  if (src_pad == nullptr) {
    delete ctx;
    throw std::runtime_error(
        "nvmultistreamtiler has no src pad for the FPS overlay probe.");
  }
  gst_pad_add_probe(
      src_pad, GST_PAD_PROBE_TYPE_BUFFER, overlay_probe, ctx,
      [](gpointer data) { delete static_cast<OverlayCtx*>(data); });
  gst_object_unref(src_pad);
}

}  // namespace mcrt
